### DISCUSSION TITLE
Elysia 1.1 Feedback

### DISCUSSION THREAD

**Discussion by SaltyAom** [elysiajs/elysia#893] [Ideas]
Hi, same as usual.

We have begin the development of the next version of Elysia.

So we would like to hear your thoughts, how was your days with Elysia going?
You can suggest anything from proposing a new feature, what you feel like is missing, or just sharing your experience.

Anyway, hope you have a nice day, thanks.

![Don Quixote dancing](https://github.com/user-attachments/assets/1fbd6fe1-9989-4315-8810-0f83d1971be4)

**Comment by gsemyong** [comment]
I am in love with Elysia! Here are some features I would like to see:

1. Exposing cookie parser to be used outside route handlers
2. Making possible one-click to definition from treaty client to route handler
3. Integration with tanstack query for eden
4. Hosting SSR frameworks like remix (I know there is a 3rd party plugin that does this, but it doesn't seem to be either production ready or moving much)
5. More bun init-like default tsconfig when running `bun create elysia app`

I am super grateful to all the people working on this project and can't wait to see what's coming :)

**Reply by Blankeos** [reply]
on Hosting SSR Frameworks, have you checked out https://vike.dev yet? So far it's really good for Vue, React, and SolidJS.
Try checking this: https://github.com/Ryooshuu/vike-elysiajs-sample

**Reply by MatthewAry** [reply]
I have been feeling the pain of 1, 2, and 3. 🤞

**Reply by kravetsone** [reply]
> I am in love with Elysia! Here are some features I would like to see:
> 
> 1. Exposing cookie parser to be used outside route handlers
> 2. Making possible one-click to definition from treaty client to route handler
> 3. Integration with tanstack query for eden
> 4. Hosting SSR frameworks like remix (I know there is a 3rd party plugin that does this, but it doesn't seem to be either production ready or moving much)
> 5. More bun init-like default tsconfig when running `bun create elysia app`
> 
> I am super grateful to all the people working on this project and can't wait to see what's coming :)
> 
> 
> 

U can use `bun create elysiajs app` instead
Also if u want to improve remix plugin feels free to send PR

**Reply by gsemyong** [reply]
> on Hosting SSR Frameworks, have you checked out https://vike.dev yet? So far it's really good for Vue, React, and SolidJS. Try checking this: https://github.com/Ryooshuu/vike-elysiajs-sample

I will explore it for sure! Thanks :)

**Comment by Bewinxed** [comment]
My christmas wish is auto inferred return type in endpoints without me having to write the schema for it.

**Comment by sinanusluer** [comment]
1. Integrated queue operations (bullmq etc.) 
2. Event, Listener 🫠

**Reply by kravetsone** [reply]
Queue shouldn't be in elysia codebase
Great bullmq already exists so no need to craft new one elysia backed

I also create wrapper for it https://github.com/kravetsone/jobify

**Reply by sinanusluer** [reply]
I like your work and logic for Bullmq. Thank you, I will look into it 🤟

**Reply by gsemyong** [reply]
I would also recommend you checking out inngest. It can be self hosted as a single container and provides durable execution, crons, etc. with typesafe sdk, ui and cli :)

**Reply by kravetsone** [reply]
> I would also recommend you checking out inngest. It can be self hosted as a single container and provides durable execution, crons, etc. with typesafe sdk, ui and cli :)

can u share links?

**Reply by kravetsone** [reply]
> I like your work and logic for Bullmq. Thank you, I will look into it 🤟

wdym by 2.?

**Reply by gsemyong** [reply]
> > I would also recommend you checking out inngest. It can be self hosted as a single container and provides durable execution, crons, etc. with typesafe sdk, ui and cli :)
> 
> can u share links?

Sure, here are some relevant links:
https://www.inngest.com/
https://www.inngest.com/docs/self-hosting

**Reply by kravetsone** [reply]
> > > I would also recommend you checking out inngest. It can be self hosted as a single container and provides durable execution, crons, etc. with typesafe sdk, ui and cli :)
> > 
> > can u share links?
> 
> Sure, here are some relevant links:
> https://www.inngest.com/
> https://www.inngest.com/docs/self-hosting

Looks interesting 
What about pricing? Didn't find info about monetization

**Reply by kravetsone** [reply]
> > > > I would also recommend you checking out inngest. It can be self hosted as a single container and provides durable execution, crons, etc. with typesafe sdk, ui and cli :)
> > > 
> > > can u share links?
> > 
> > Sure, here are some relevant links:
> > https://www.inngest.com/
> > https://www.inngest.com/docs/self-hosting
> 
> Looks interesting 
> What about pricing? Didn't find info about monetization 

Oh i see in second link
Looks fine

**Comment by Blankeos** [comment]
Stable node support would be bonkers good. 🎯

**Comment by hansaskov** [comment]
I’d love to see inferred error types directly from schema definitions in Eden Treaty. Specifically, it would be helpful to set custom error messages within a schema—like `"Please provide a valid email"` for an email field—and then see that exact message returned on the client side when handling errors. Currently, the error type returns as `unknown`, which makes handling these cases more challenging.

Here’s a test case to illustrate this idea:

```typescript
import { describe, expect, it } from 'bun:test';
import { treaty } from "@elysiajs/eden";
import { Elysia, t } from "elysia";

const app = new Elysia()
    .post("/email", ({ body }) => `${body.email}`, {
        body: t.Object({
            email: t.String({
                format: "email",
                error: "Please provide a valid email"
            })
        })
    });

const api = treaty(app);

describe('Eden Treaty error handling', () => {
    it("Correct email input", async () => {
        const { data, error } = await api.email.post({ email: "hans@askov.dk" });
        expect(data).toBe("hans@askov.dk");
    });

    it("Incorrect email input", async () => {
        const { data, error } = await api.email.post({ email: "not an email" });
        expect(error?.status).toBe(422);
        expect(error?.value).toBe("Please provide a valid email");
    });
});

/*
Expected Error Type:
const error: {
    status: 422;
    value: string;
} | null;

Actual Error Type:
const error: {
    status: unknown;
    value: unknown;
} | null;
*/
```

### Explanation
In the example above:
When a user provides an invalid email (e.g., `"not an email"`), the schema’s custom error message `"Please provide a valid email"` is set, and we’d expect it to appear in the error response. However, instead of the expected error type with `status` and `value`, we currently get an error type with both fields as `unknown`.

Having the custom error messages carry over to the client would improve error handling consistency and make applications using Eden Treaty more user-friendly. 

Note: This might be a skill issue on my part, so please enlighten me if i am mistaken.

Thank you for considering this suggestion!

**Reply by crishoj** [reply]
@hansaskov — running your example locally, it works as you expect it to:

<img width="626" alt="image" src="https://github.com/user-attachments/assets/6ef854d5-a777-4049-9f73-208b9108f1db">

The client receives an error with status `422` and value `Please provide a valid email`.

**Reply by hansaskov** [reply]
Hey crishoj. I should have been more clear from my explanation. At runtime the test works as expected, but the Typescript type for the "error" object given is "unknown" Where it would be nice if it was inferred as a string.

**Reply by crishoj** [reply]
Ah, you mean the properties in the `error` object?

<img width="605" alt="image" src="https://github.com/user-attachments/assets/d930e7d1-7a53-4eb2-8ba5-a1de0c77a164">

I guess it would be nice if they were inferred as `number` and `string`, respectively.

**Comment by relevantsam** [comment]
Integrated compression - the community extension is no longer functional and gzipped responses seems like a sensible performance feature to have baked in.

**Reply by kravetsone** [reply]
I guess it can be official plugin

No need to do it out of the box

**Reply by relevantsam** [reply]
Maybe? Seems like an important feature for performant web applications.

**Reply by kravetsone** [reply]
> Maybe? Seems like an important feature for performant web applications.

every great production application use reverse proxy such as nginx which have more perfomant operations for gzipping and etc

so no need to do it built-in. it is not scope of elysia

but first-of-class plugin support would be fine
just

```ts
const app = new Elysia().use(compression())
```

and all peoples would be happy

**Comment by BleedingDev** [comment]
Allow more than just Typebox. I would love to see #EffectTS. ❤️

**Comment by rhuanbarreto** [comment]
The journey has been good so far with some bumps in the road about Eden. Some desires for 1.1:

1. First class support to Date types. There are 2 PRs in Eden trying to help in this matter. Today you can't just use t.Date(). We had to do our own `bun patch` of Eden locally because PRs don't even get reviewed/commented by the maintainers.
2. Better support for errors in Eden, like raising errors instead of having to inspect the error property of the response. This would help a lot on integrating with react query once errors would be treated in the library.
3. More streamlined client for websockets. Today it's very quirky.
4. Interop with other languages. An Eden treaty for Python with exported type definitions would help a lot into integrating codebases in different languages with full e2e type safety. Like swagger generator but elysia style

**Reply by crishoj** [reply]
Feedback/review on open PRs would indeed be very welcome.

**Comment by rempas** [comment]
I would to see an addition to documentation about how to use [Preact](https://preactjs.com/) and a template to `bun create elysia`  to use it by default, with an example.

**Comment by rolandszpond** [comment]
I would like to see cache-control get an update. When using `set.headers['cache-control'] = 'max-age=3600'` I expect it to be able to cache on the users browser, but it seems like it just caches on a random Cloudflare endpoint and never caching on the browser? If that's the case, I would rather `max-age` be used to handle the cache on the users browser, and setting an additional options `s-maxage=3600` to be used to cache on a CDN, similar to how Firebase handles the caching on both browser and CDN. Additionally, I think it would be useful to have control of the CDN, like be able to set my own Cloudflare to use instead of what is seems to be using is some random account? and finally, an easy way to purge the CDN cache, maybe a `purge` settings being passed to the handler, or something. More info on the Firebase thing: https://firebase.google.com/docs/hosting/manage-cache#set_cache-control

**Reply by rolandszpond** [reply]
Oh, almost forgot another wishlist thing I want, I see examples of integrations with frameworks like Nextjs, but I can't seem to find a way to run Elysia on the same bun server as Nuxtjs while using their server paths. The only solution I've found so far is having to two different servers at the same time.

**Comment by crishoj** [comment]
Would love to have auto-reconnection in the Eden WebSocket client.

I ended up using https://vueuse.org/core/useWebSocket/#auto-reconnection as WS client instead, but would much prefer to use Eden.

**Comment by crishoj** [comment]
Regarding `aot`:
- Either fully support `{ aot: false }` with comprehensive test coverage.
- Or explicitly deprecate the `{ aot: false }` option.

**Comment by MatthewAry** [comment]
Would love to see a solution for tacking down anonymous functions in traces. Been trying to track down a perf issue and I'm struggling.

**Comment by crishoj** [comment]
A convenient way to inspect/debug optimized handlers could be nice.

Currently, optimized handler code is bit of a black box. Perhaps an option to write the generated code to disk or log to console.

**Comment by sebastiansandqvist** [comment]
Tanstack Query support within Eden, including adapters for each clientside library that they support (or at least Solid/Svelte/React) would be pretty great!

**Comment by crishoj** [comment]
Regarding developer experience with Elysia, it has been generally fantastic.

One detail really caught me off guard, though: The fact that [response normalization](https://elysiajs.com/essential/validation#normalize) recursively _mutates_ values used in responses with `delete obj[property]` 😨

This resulted in strange bugs with properties that suddenly became `undefined` in long-lived service objects.

Reported as #897, and solution proposed as #905. Would love to get team feedback on this 🙏

**Comment by myodan** [comment]
I think we need more example projects, especially for monorepo, ex) turborepo monorepo, bun workspace mono repo, jwt, session, common project structures etc...

**Reply by hansaskov** [reply]
I recently created a monorepo with Bun, Elysia and Svelte. Feel free to check it out and get inspiration for your needs
https://github.com/hansaskov/monorepo-elysia-svelte
